File: lua/custom_plugins/auditscope/mind/report.lua
local Path = require("plenary.path")
local db = require("custom_plugins.auditscope.mind.db")

local M = {}

local function format_time(ts)
  if not ts then
    return "unknown"
  end
  return os.date("%Y-%m-%d %H:%M", ts)
end

local function sanitize_text(text)
  if not text then
    return ""
  end
  local clean = tostring(text):gsub("[\r\n]+", " ")
  return clean
end

local function safe_filename(value)
  if value == nil then
    return nil
  end
  local name = tostring(value)
  if name == "" then
    return nil
  end
  name = name:gsub("[/\\]", "_")
  name = name:gsub("%s+", "_")
  name = name:gsub("[^%w%._-]", "_")
  name = name:gsub("_+", "_")
  name = name:gsub("^[_%.]+", "")
  name = name:gsub("[_%.]+$", "")
  if name == "" then
    return nil
  end
  return name
end

local function relative_path(file, root)
  if not file or not root then
    return file
  end
  local normalized_root = vim.fs.normalize(root)
  local normalized_file = vim.fs.normalize(file)
  local prefix = normalized_root .. "/"
  if vim.startswith(normalized_file, prefix) then
    return normalized_file:sub(#prefix + 1)
  end
  return normalized_file
end

local function format_line_range(start_line, end_line)
  if not start_line then
    return ""
  end
  if not end_line or start_line == end_line then
    return tostring(start_line)
  end
  return string.format("%d-%d", start_line, end_line)
end

local function group_nodes(nodes)
  local groups = {}
  for _, node in ipairs(nodes) do
    local t = node.type or "note"
    groups[t] = groups[t] or {}
    table.insert(groups[t], node)
  end
  return groups
end

local function format_reference(node)
  if not node.file or not node.start_line then
    return nil
  end
  local rel = relative_path(node.file, node.repo_root)
  local range = format_line_range(node.start_line, node.end_line)
  local commit = node.commit and ("@" .. node.commit) or ""
  return string.format("`%s:%s%s`", rel or "unknown", range, commit)
end

local function format_repo_label(node)
  local repo = node.repo_name or (node.repo_root and vim.fn.fnamemodify(node.repo_root, ":t")) or nil
  if repo then
    return repo
  end
  return nil
end

function M.generate(opts)
  opts = opts or {}

  local subject = db.get_subject()
  if not subject then
    vim.notify("AuditScope: No active subject. Use :AuditSubjectNew or :AuditSubjectSelect.", vim.log.levels.ERROR)
    return nil
  end

  local nodes = db.get_nodes()
  local storage = db.get_storage_paths()
  local reports_root = storage and storage.reports_dir
  if not reports_root or reports_root == "" then
    reports_root = vim.fs.joinpath(vim.fn.stdpath("data"), "auditscope", "reports")
  end

  local report_dir = Path:new(reports_root)
  if not report_dir:exists() then
    report_dir:mkdir({ parents = true })
  end

  local report_name = safe_filename(subject.id) or safe_filename(subject.title) or ("subject_" .. os.time())
  local report_path = Path:new(vim.fs.joinpath(report_dir:absolute(), report_name .. ".md"))
  if report_path:exists() and report_path:is_dir() then
    report_name = report_name .. "_" .. os.time()
    report_path = Path:new(vim.fs.joinpath(report_dir:absolute(), report_name .. ".md"))
  end

  local lines = {}
  table.insert(lines, "# " .. (subject.title or "Untitled Subject"))
  table.insert(lines, "")
  table.insert(lines, string.format("- Status: %s", subject.status or "active"))
  if subject.scope and subject.scope ~= "" then
    table.insert(lines, string.format("- Scope: %s", subject.scope))
  end
  table.insert(lines, string.format("- Created: %s", format_time(subject.created_at)))
  table.insert(lines, string.format("- Updated: %s", format_time(subject.updated_at)))
  table.insert(lines, "")

  table.insert(lines, "## Executive Summary")
  table.insert(lines, "")
  if subject.summary and subject.summary ~= "" then
    table.insert(lines, sanitize_text(subject.summary))
  else
    table.insert(lines, "_No summary yet._")
  end
  table.insert(lines, "")

  local groups = group_nodes(nodes)

  local function add_group_list(title, types, prefix)
    local items = {}
    for _, t in ipairs(types) do
      if groups[t] then
        for _, n in ipairs(groups[t]) do
          table.insert(items, n)
        end
      end
    end
    if #items == 0 then
      return
    end
    table.insert(lines, "## " .. title)
    table.insert(lines, "")
    local idx = 0
    for _, n in ipairs(items) do
      idx = idx + 1
      local text = sanitize_text(n.text or "")
      if text == "" then
        text = "(empty)"
      end
      local label = prefix and string.format("%s%d", prefix, idx) or tostring(idx)
      table.insert(lines, string.format("- [%s] %s", label, text))
      local fields = {}
      local repo_label = format_repo_label(n)
      if repo_label then
        table.insert(fields, "Repo: " .. repo_label)
      end
      local ref = format_reference(n)
      if ref then
        table.insert(fields, "Location: " .. ref)
      end
      if #fields > 0 then
        table.insert(lines, "  " .. table.concat(fields, " | "))
      end
    end
    table.insert(lines, "")
  end

  local function add_group_heading(title, types, prefix)
    local items = {}
    for _, t in ipairs(types) do
      if groups[t] then
        for _, n in ipairs(groups[t]) do
          table.insert(items, n)
        end
      end
    end
    if #items == 0 then
      return
    end
    table.insert(lines, "## " .. title)
    table.insert(lines, "")
    local idx = 0
    for _, n in ipairs(items) do
      idx = idx + 1
      local text = sanitize_text(n.text or "")
      if text == "" then
        text = "(empty)"
      end
      local label = prefix and string.format("%s%d", prefix, idx) or tostring(idx)
      table.insert(lines, string.format("### [%s] %s", label, text))
      local fields = {}
      local repo_label = format_repo_label(n)
      if repo_label then
        table.insert(fields, "Repo: " .. repo_label)
      end
      local ref = format_reference(n)
      if ref then
        table.insert(fields, "Location: " .. ref)
      end
      if #fields > 0 then
        table.insert(lines, table.concat(fields, " | "))
      end
      table.insert(lines, "")
    end
  end

  add_group_list("Evidence", { "evidence", "fact" }, "E")
  add_group_list("Notes", { "note" }, "N")
  add_group_list("Insights", { "insight" }, "I")
  add_group_list("Assumptions", { "assumption" }, "A")
  add_group_list("Invariants", { "invariant" }, "V")
  add_group_list("Open Questions", { "question" }, "Q")
  add_group_list("Hypotheses", { "hypothesis" }, "H")
  add_group_heading("Findings", { "finding" }, "F")
  add_group_heading("Decisions", { "decision" }, "D")
  add_group_heading("Risks", { "risk" }, "R")

  report_path:write(table.concat(lines, "\n"), "w")

  if opts.open then
    vim.cmd("edit " .. report_path:absolute())
  end

  vim.notify(string.format("AuditScope: Report generated at %s", report_path:absolute()), vim.log.levels.INFO)
  return report_path:absolute()
end

return M

File: lua/custom_plugins/auditscope/cli/internal/report/report.go
package report

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/kobayakawayuu/auditscope/pkg/models"
)

type Generator struct {
	Subject    *models.Subject
	Nodes      []models.Node
	Edges      []models.Edge
	ReportsDir string
}

func NewGenerator(subject *models.Subject, nodes []models.Node, edges []models.Edge, reportsDir string) *Generator {
	return &Generator{
		Subject:    subject,
		Nodes:      nodes,
		Edges:      edges,
		ReportsDir: reportsDir,
	}
}

func (g *Generator) getRepoRoot() string {
	if g.Subject != nil && g.Subject.RepoRoot != "" {
		return g.Subject.RepoRoot
	}
	for _, n := range g.Nodes {
		if n.RepoRoot != "" {
			return n.RepoRoot
		}
	}
	return ""
}

func (g *Generator) resolveFilePath(file string) string {
	if file == "" {
		return ""
	}
	if filepath.IsAbs(file) {
		return file
	}
	repoRoot := g.getRepoRoot()
	if repoRoot != "" {
		return filepath.Join(repoRoot, file)
	}
	return file
}

func (g *Generator) readFileLines(filePath string, startLine, endLine int) []string {
	if filePath == "" || startLine == 0 {
		return nil
	}

	data, err := os.ReadFile(filePath)
	if err != nil {
		return nil
	}

	lines := strings.Split(string(data), "\n")
	if startLine < 1 {
		startLine = 1
	}
	if endLine == 0 || endLine > len(lines) {
		endLine = len(lines)
	}
	if startLine > len(lines) {
		return nil
	}

	startIdx := startLine - 1
	endIdx := endLine
	if endIdx > len(lines) {
		endIdx = len(lines)
	}

	return lines[startIdx:endIdx]
}

func (g *Generator) detectLanguage(file string) string {
	if file == "" {
		return ""
	}
	ext := strings.ToLower(filepath.Ext(file))
	langMap := map[string]string{
		".rs":    "rust",
		".sol":   "solidity",
		".go":    "go",
		".ts":    "typescript",
		".tsx":   "typescript",
		".js":    "javascript",
		".jsx":   "javascript",
		".py":    "python",
		".c":     "c",
		".cpp":   "cpp",
		".h":     "c",
		".hpp":   "cpp",
		".java":  "java",
		".kt":    "kotlin",
		".swift": "swift",
		".rb":    "ruby",
		".lua":   "lua",
		".sh":    "bash",
		".yaml":  "yaml",
		".yml":   "yaml",
		".json":  "json",
		".toml":  "toml",
		".md":    "markdown",
	}
	if lang, ok := langMap[ext]; ok {
		return lang
	}
	return ""
}

func (g *Generator) renderCodeBlock(file string, startLine, endLine int) []string {
	filePath := g.resolveFilePath(file)
	lines := g.readFileLines(filePath, startLine, endLine)
	if len(lines) == 0 {
		return nil
	}

	lang := g.detectLanguage(file)
	var result []string
	result = append(result, fmt.Sprintf("```%s", lang))

	lineRange := fmt.Sprintf("%d", startLine)
	if endLine > 0 && endLine != startLine {
		lineRange = fmt.Sprintf("%d-%d", startLine, endLine)
	}
	relPath := file
	repoRoot := g.getRepoRoot()
	if repoRoot != "" && strings.HasPrefix(file, repoRoot) {
		relPath = strings.TrimPrefix(file, repoRoot)
		if relPath[0] == '/' {
			relPath = relPath[1:]
		}
	}
	result = append(result, fmt.Sprintf("/// %s:%s", relPath, lineRange))

	for i, line := range lines {
		lineNum := startLine + i
		result = append(result, fmt.Sprintf("%d| %s", lineNum, line))
	}

	result = append(result, "```")
	return result
}

func (g *Generator) Generate() (string, error) {
	if g.Subject == nil {
		return "", fmt.Errorf("no subject")
	}

	if err := os.MkdirAll(g.ReportsDir, 0o755); err != nil {
		return "", err
	}

	reportName := safeFilename(g.Subject.ID)
	if reportName == "" {
		reportName = safeFilename(g.Subject.Title)
	}
	if reportName == "" {
		reportName = fmt.Sprintf("subject_%d", time.Now().Unix())
	}

	reportPath := filepath.Join(g.ReportsDir, reportName+".md")

	var lines []string

	lines = append(lines, fmt.Sprintf("# %s", g.Subject.Title))
	lines = append(lines, "")
	lines = append(lines, fmt.Sprintf("- Status: %s", g.Subject.Status))
	if g.Subject.Scope != "" {
		lines = append(lines, fmt.Sprintf("- Scope: %s", g.Subject.Scope))
	}
	lines = append(lines, fmt.Sprintf("- Created: %s", formatTime(g.Subject.CreatedAt)))
	lines = append(lines, fmt.Sprintf("- Updated: %s", formatTime(g.Subject.UpdatedAt)))
	lines = append(lines, "")

	lines = append(lines, "## Executive Summary")
	lines = append(lines, "")
	if g.Subject.Summary != "" {
		lines = append(lines, sanitizeText(g.Subject.Summary))
	} else {
		lines = append(lines, "_No summary yet._")
	}
	lines = append(lines, "")

	g.addGroupList(&lines, "Evidence", []string{models.NodeTypeEvidence, models.NodeTypeFact}, "E")
	g.addGroupList(&lines, "Notes", []string{models.NodeTypeNote}, "N")
	g.addGroupList(&lines, "Insights", []string{models.NodeTypeInsight}, "I")
	g.addGroupList(&lines, "Assumptions", []string{models.NodeTypeAssumption}, "A")
	g.addGroupList(&lines, "Invariants", []string{models.NodeTypeInvariant}, "V")
	g.addGroupList(&lines, "Open Questions", []string{models.NodeTypeQuestion}, "Q")
	g.addGroupList(&lines, "Hypotheses", []string{models.NodeTypeHypothesis}, "H")
	g.addGroupHeading(&lines, "Findings", []string{models.NodeTypeFinding}, "F")
	g.addGroupHeading(&lines, "Decisions", []string{models.NodeTypeDecision}, "D")
	g.addGroupHeading(&lines, "Risks", []string{models.NodeTypeRisk}, "R")

	content := strings.Join(lines, "\n")
	if err := os.WriteFile(reportPath, []byte(content), 0o644); err != nil {
		return "", err
	}

	return reportPath, nil
}

func (g *Generator) groupNodes() map[string][]models.Node {
	groups := make(map[string][]models.Node)
	for _, n := range g.Nodes {
		t := n.Type
		if t == "" {
			t = models.NodeTypeNote
		}
		groups[t] = append(groups[t], n)
	}
	return groups
}

func (g *Generator) addGroupList(lines *[]string, title string, types []string, prefix string) {
	var items []models.Node
	for _, t := range types {
		items = append(items, g.groupNodes()[t]...)
	}
	if len(items) == 0 {
		return
	}

	*lines = append(*lines, fmt.Sprintf("## %s", title))
	*lines = append(*lines, "")

	for i, n := range items {
		title := sanitizeText(n.GetTitle())
		if title == "" {
			title = "(empty)"
		}
		label := fmt.Sprintf("%s%d", prefix, i)
		*lines = append(*lines, fmt.Sprintf("- [%s] %s", label, title))

		if n.Description != "" {
			*lines = append(*lines, fmt.Sprintf("  %s", sanitizeText(n.Description)))
		}

		var fields []string
		if n.RepoName != "" {
			fields = append(fields, fmt.Sprintf("Repo: %s", n.RepoName))
		}
		if n.File != "" && n.StartLine > 0 {
			ref := formatReference(n)
			if ref != "" {
				fields = append(fields, fmt.Sprintf("Location: %s", ref))
			}
		}
		if len(fields) > 0 {
			*lines = append(*lines, fmt.Sprintf("  %s", strings.Join(fields, " | ")))
		}

		snippets := collectSnippets(n)
		if len(snippets) > 0 {
			*lines = append(*lines, "  Snippets:")
			for _, s := range snippets {
				snippetLines := strings.Split(s.Text, "\n")
				for _, l := range snippetLines {
					*lines = append(*lines, fmt.Sprintf("    %s", l))
				}
				*lines = append(*lines, "  ")
			}
		}
	}
	*lines = append(*lines, "")
}

func (g *Generator) addGroupHeading(lines *[]string, title string, types []string, prefix string) {
	var items []models.Node
	for _, t := range types {
		items = append(items, g.groupNodes()[t]...)
	}
	if len(items) == 0 {
		return
	}

	*lines = append(*lines, fmt.Sprintf("## %s", title))
	*lines = append(*lines, "")

	for i, n := range items {
		title := sanitizeText(n.GetTitle())
		if title == "" {
			title = "(empty)"
		}
		label := fmt.Sprintf("%s%d", prefix, i)
		*lines = append(*lines, fmt.Sprintf("### [%s] %s", label, title))

		if n.Description != "" {
			*lines = append(*lines, sanitizeText(n.Description))
		}

		var fields []string
		fields = append(fields, fmt.Sprintf("\n> "))
		if n.RepoName != "" {
			fields = append(fields, fmt.Sprintf("Repo: %s", n.RepoName))
		}
		if n.File != "" && n.StartLine > 0 {
			ref := formatReference(n)
			if ref != "" {
				fields = append(fields, fmt.Sprintf("Location: %s", ref))
			}
		}
		if len(fields) > 0 {
			*lines = append(*lines, strings.Join(fields, " | "))
		}

		if n.File != "" && n.StartLine > 0 {
			codeLines := g.renderCodeBlock(n.File, n.StartLine, n.EndLine)
			if len(codeLines) > 0 {
				*lines = append(*lines, "")
				*lines = append(*lines, codeLines...)
			}
		}

		snippets := collectSnippets(n)
		if len(snippets) > 0 {
			*lines = append(*lines, "")
			*lines = append(*lines, "**Additional Snippets:**")
			for _, s := range snippets {
				lang := g.detectLanguage(s.File)
				if lang != "" {
					*lines = append(*lines, fmt.Sprintf("```%s", lang))
				} else {
					*lines = append(*lines, "```")
				}
				*lines = append(*lines, strings.Split(s.Text, "\n")...)
				*lines = append(*lines, "```")
			}
		}
		*lines = append(*lines, "")
	}
}

func formatTime(ts int64) string {
	if ts == 0 {
		return "unknown"
	}
	return time.Unix(ts, 0).Format("2006-01-02 15:04")
}

func sanitizeText(text string) string {
	if text == "" {
		return ""
	}
	return text
	// return strings.ReplaceAll(strings.ReplaceAll(text, "\r", " "), "\n", " ")
}

func safeFilename(value string) string {
	if value == "" {
		return ""
	}
	name := value
	name = strings.Map(func(r rune) rune {
		if r == '/' || r == '\\' || r == ' ' {
			return '_'
		}
		if (r >= 'a' && r <= 'z') || (r >= 'A' && r <= 'Z') || (r >= '0' && r <= '9') || r == '_' || r == '-' || r == '.' {
			return r
		}
		return '_'
	}, name)
	name = strings.Trim(name, "_.")
	return name
}

func formatReference(n models.Node) string {
	if n.File == "" || n.StartLine == 0 {
		return ""
	}
	rel := n.File
	if n.RepoRoot != "" {
		rel = strings.TrimPrefix(n.File, n.RepoRoot+"/")
	}
	lineRange := fmt.Sprintf("%d", n.StartLine)
	if n.EndLine > 0 && n.EndLine != n.StartLine {
		lineRange = fmt.Sprintf("%d-%d", n.StartLine, n.EndLine)
	}
	commit := ""
	if n.Commit != "" {
		commit = "@" + n.Commit
	}
	return fmt.Sprintf("`%s:%s%s`", rel, lineRange, commit)
}

func collectSnippets(n models.Node) []models.CodeSnippet {
	var snippets []models.CodeSnippet
	if len(n.CodeSnippets) > 0 {
		for _, s := range n.CodeSnippets {
			if s.Text != "" {
				snippets = append(snippets, s)
			}
		}
	}
	if len(snippets) == 0 && n.CodeSnippet != "" {
		snippets = append(snippets, models.CodeSnippet{Text: n.CodeSnippet})
	}
	return snippets
}
